File: 백준/Silver/2343. 기타 레슨/기타 레슨.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, m;
	static int[] arr;

	static int answer;

	public static void main(String[] args) throws Exception {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(bf.readLine());

		n = Integer.parseInt(st.nextToken()); // 강의의 수
		m = Integer.parseInt(st.nextToken()); // 블루레이 개수

		arr = new int[n];
		answer = 0;

		// start와 end는 각 블루레이의 크기를 정하기 위함
		int start = 0;
		int end = 0;

		st = new StringTokenizer(bf.readLine());
		for (int i = 0; i < n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
			end += arr[i];
			start = Math.max(start, arr[i]);
		}

		int mid = 0;
		// 45 + 9 /2 = 27

		while (start < end) {
			mid = (start + end) / 2; // 블루레이의 크기

			int sum = 0;
			int cnt = 1;
			for (int i = 0; i < n; i++) {
				sum += arr[i];
				if (sum > mid) {
					sum = arr[i];
					cnt++;
				}
			}
			if (cnt > m) {
				start = mid + 1;
			} else {
				end = mid;
			}
		}

		System.out.println(end);
	}

}

File: 백준/Silver/2343. 기타 레슨/기타 레슨.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? './dev/stdin' : 'index.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const [N, M] = input[0].split(' ').map(Number);
const arr = input[1].split(' ').map(Number);

let left = Math.max(...arr);
let right = arr.reduce((acc, cur) => acc + cur, 0);
let answer = right;

while (left <= right) {
  let mid = Math.floor((left + right) / 2);

  let count = 1;
  let sum = 0;
  for (const v of arr) {
    sum += v;
    if (sum > mid) {
      sum = v;
      count++;
    }
  }

  if (count > M) {
    left = mid + 1;
  } else {
    right = mid - 1;
    answer = Math.min(answer, mid);
  }
}

console.log(answer);